### PR TITLE
This PR Addresses MYRIAD-153: tasks not finishing when FGS is enabled…

### PR DESCRIPTION
**Pull request by DarinJ**
….  The root cause is that when a container is preempted by the Resource Manager, it can go from AQUIRED to RELEASED in which case the Mesos task would be started, however since the container never got to state RUNNING the `ContainerManagerImpl` never called `MyriadAuxService.initializeContainer` or `MyriadAuxService.stopContainer` the solution is to intercept the method `releaseContainers` in `AbstractScheduler` to kill any yarn_task which is preempted this way and immediately adjust the resources of the node so another container doesn't start expecting these resources.

### CODE REVIEW

**Comment by DarinJ** [comment]
@smarella @sdaingade @yufeldman This is WIP but I'm pretty sure it's the right way to current it.  I'll be cleaning up the code over the next day or two, but if you could take a look before the sync Wed we can discuss.

**Comment by DarinJ** [comment]
@kensipe @smarella @sdaingade Any comments on this, I've tested pretty thoroughly and would like to merge.

**Comment by smarella** [comment]
@DarinJ - sorry for the delay. Will take a look at this today.

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java [hunk 1]
```diff
@@ -98,6 +105,45 @@ public boolean allowCallBacksForNode(NodeId nodeManager) {
...
+    if (type.equals(RMContainerEventType.KILL) || type.equals(RMContainerEventType.RELEASED)) {
+      LOGGER.info("{} completed with exit status {}, killing cooresponding mesos task.", rmContainer.getContainerId().toString(), type);
+      myKillContainer(rmContainer);
+    }
+  }
+
+  private synchronized void myKillContainer(RMContainer rmContainer) {
+    if (rmContainer != null && rmContainer.getContainer() != null) {
+      Protos.TaskID taskId = containerToTaskId(rmContainer);
+      //TODO (darinj) Reliable messaging
```
Comment (smarella) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: Is this TODO still valid? Is ```myriadDriver.kill(taskId)``` unreliable?

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java [hunk 2]
```diff
@@ -98,6 +105,45 @@ public boolean allowCallBacksForNode(NodeId nodeManager) {
...
+
+  @Override
+  public void beforeCompletedContainer(RMContainer rmContainer, ContainerStatus containerStatus, RMContainerEventType type) {
+    if (type.equals(RMContainerEventType.KILL) || type.equals(RMContainerEventType.RELEASED)) {
+      LOGGER.info("{} completed with exit status {}, killing cooresponding mesos task.", rmContainer.getContainerId().toString(), type);
+      myKillContainer(rmContainer);
+    }
+  }
+
+  private synchronized void myKillContainer(RMContainer rmContainer) {
```
Comment (smarella) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: Can you please pick a better name for this function?

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java [hunk 3]
```diff
@@ -98,6 +105,45 @@ public boolean allowCallBacksForNode(NodeId nodeManager) {
...
+      Protos.TaskID taskId = containerToTaskId(rmContainer);
+      //TODO (darinj) Reliable messaging
+      state.makeTaskKillable(taskId);
+      myriadDriver.kill(taskId);
+      String hostname = rmContainer.getContainer().getNodeId().getHost();
+      Node node = nodeStore.getNode(hostname);
+      if (node != null) {
+        RMNode rmNode = node.getNode().getRMNode();
+        Resource resource = rmContainer.getContainer().getResource();
+        Resource diff = ResourceUtils.componentwiseMax(ZERO_RESOURCE, Resources.subtract(rmNode.getTotalCapability(), resource));
```
Comment (smarella) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: Why is this required? Do you see any danger of ```diff``` going negative?

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java [hunk 4]
```diff
@@ -98,6 +105,45 @@ public boolean allowCallBacksForNode(NodeId nodeManager) {
...
+        Resources.subtract(rmNode.getTotalCapability(), resource);
```
Comment (smarella) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: This is perhaps redundant as you are ignoring the return value of ```Resources.subtract()```

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java [hunk 5]
```diff
@@ -196,13 +242,16 @@ protected void handleContainerAllocation(RMNode rmNode) {
...
   public void setNodeCapacity(RMNode rmNode, Resource newCapacity) {
-    rmNode.getTotalCapability().setMemory(newCapacity.getMemory());
-    rmNode.getTotalCapability().setVirtualCores(newCapacity.getVirtualCores());
-    LOGGER.debug("Setting capacity for node {} to {}", rmNode.getHostName(), newCapacity);
-    // updates the scheduler with the new capacity for the NM.
-    // the event is handled by the scheduler asynchronously
-    rmContext.getDispatcher().getEventHandler().handle(new NodeResourceUpdateSchedulerEvent(rmNode, ResourceOption.newInstance(
-        rmNode.getTotalCapability(), RMNode.OVER_COMMIT_TIMEOUT_MILLIS_DEFAULT)));
+    //NOOP prevent YARN warning changing to same size
+    if (!Resources.equals(rmNode.getTotalCapability(), newCapacity)) {
```
Comment (smarella) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: Pretty cool!

File: myriad-executor/src/main/java/org/apache/myriad/executor/MyriadExecutor.java [hunk 6]
```diff
@@ -70,10 +70,11 @@ public void launchTask(final ExecutorDriver driver, final TaskInfo task) {
 
   @Override
   public void killTask(ExecutorDriver driver, TaskID taskId) {
-    LOGGER.debug("killTask received for taskId: " + taskId.getValue());
+    String taskIdString = taskId.getValue();
+    LOGGER.debug("killTask received for taskId: " + taskIdString);
     TaskStatus status;
 
-    if (!taskId.toString().contains(MyriadExecutorAuxService.YARN_CONTAINER_TASK_ID_PREFIX)) {
+    if (!taskIdString.contains(MyriadExecutorAuxService.YARN_CONTAINER_TASK_ID_PREFIX)) {
```
Comment (smarella) [inline myriad-executor/src/main/java/org/apache/myriad/executor/MyriadExecutor.java]: I didn't understand the need to change from ```taskId.toString()``` to ```taskId.getValue()``` and to introduce ```YARN_CONTAINER_FULL_PREFIX```. Can you please explain?

**Comment by smarella** [comment]
@DarinJ - great work :100:

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java (hunk 1 above)
Comment (DarinJ) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: It's slightly unreliable in the event of a network partition, but it doesn't matter now as the zero-size-nodemanger immediately goes to TASK_LOST.  I'm looking at the StateStore to see about adding a timeout delay for node managers for brief partitions, at that point we'll need to have a thread monitor a kill queue, while a TASK_KILLED event would remove the task from the kill queue.

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java (hunk 2 above)
Comment (DarinJ) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: Yes, I had planned to and forgot.

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java (hunk 4 above)
Comment (DarinJ) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: Yep it's a no op will delete.

File: myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java (hunk 3 above)
Comment (DarinJ) [inline myriad-scheduler/src/main/java/org/apache/myriad/scheduler/fgs/YarnNodeCapacityManager.java]: Pure defense.

File: myriad-executor/src/main/java/org/apache/myriad/executor/MyriadExecutor.java (hunk 6 above)
Comment (DarinJ) [inline myriad-executor/src/main/java/org/apache/myriad/executor/MyriadExecutor.java]: I've got no opinion on getValue vs toString, the main issue was I uses taskId.toString more than once so thought I should keep it around.  I'll set taskIdString = taskId.toString()

**Comment by DarinJ** [comment]
Merged, typo prevented asfgit from closing.  So I'm closing manually.
